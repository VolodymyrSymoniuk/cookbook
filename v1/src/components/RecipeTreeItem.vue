<template>
  <li>
    <div
      :class="[isFolder ? 'bold' : '', 'item', 'pointer']"
      @click="toggle"
      @dblclick="makeFolder"
    >
      {{ item.description }}
      <span v-if="isFolder">[{{ isOpen ? "-" : "+" }}]</span>
      <span
        v-if="!isRoot"
        @click.stop="displayRecipe"
        class="material-icons show-icon"
        :class="{ active: isRecipeActive }"
      >
        visibility
      </span>
    </div>
    <ul v-show="isOpen" v-if="isFolder" class="children">
      <RecipeTreeItem
        v-for="(childId, index) in item.children"
        :key="index"
        :recipeId="childId"
      ></RecipeTreeItem>
      <li class="pointer add-item" @click="addItem">+</li>
    </ul>
  </li>
</template>

<script>
import shortid from "shortid";
import { mapMutations, mapActions } from "vuex";
import RecipeTreeItem from "@/components/RecipeTreeItem";

export default {
  name: "RecipeTreeItem",
  components: { RecipeTreeItem },
  props: {
    recipeId: String
  },
  data: () => ({
    isOpen: false
  }),
  mounted() {
    if (this.isRoot) {
      this.isOpen = true;
    }
  },
  computed: {
    item() {
      return this.$store.getters.getRecipeById(this.recipeId);
    },
    isFolder() {
      return this.item.children && this.item.children.length;
    },
    isRoot() {
      return this.recipeId === this.$store.getters.rootRecipeId;
    },
    isRecipeActive() {
      return this.$store.getters.isRecipeActive(this.recipeId);
    }
  },
  methods: {
    ...mapMutations(["setActiveRecipeById", "makeRecipeParent"]),
    ...mapActions(["addNewRecipe"]),
    toggle() {
      if (this.isFolder) {
        this.isOpen = !this.isOpen;
      }
    },
    makeFolder() {
      if (!this.isFolder) {
        this.makeRecipeParent(this.item);
        this.addItem(this.item);
        this.isOpen = true;
      }
    },
    async addItem() {
      let { id: parent, description } = this.item;
      if (parent === "root") {
        try {
          description = await this.$prompt("Enter description of new recipe:");
        } catch (_) {
          this.$alert("Invalid recipe description entered!");
          return;
        }
      }

      this.addNewRecipe({
        id: shortid.generate(),
        description,
        versions: [],
        parent,
        createdAt: Date.now()
      });
    },
    displayRecipe() {
      this.setActiveRecipeById(this.recipeId);
    }
  }
};
</script>
