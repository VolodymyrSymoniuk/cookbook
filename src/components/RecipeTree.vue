<template>
  <section>
    <ul>
      <RecipeTreeItem :recipeId="rootRecipeId"></RecipeTreeItem>
    </ul>
  </section>
</template>

<script>
import RecipeTreeItem from "@/components/RecipeTreeItem";
import { mapGetters } from "vuex";

export default {
  name: "RecipeTree",
  components: { RecipeTreeItem },
  mounted() {
    this.$store.dispatch("fetchRecipes");
  },
  computed: {
    ...mapGetters(["rootRecipeId"])
  }
};
</script>
