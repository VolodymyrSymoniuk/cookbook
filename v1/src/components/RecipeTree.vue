<template>
  <section>
    <ul>
      <RecipeTreeItem :recipeId="rootRecipeId" />
    </ul>
  </section>
</template>

<script>
import RecipeTreeItem from "@/components/RecipeTreeItem";
import { mapGetters } from "vuex";

export default {
  name: "RecipeTree",
  components: { RecipeTreeItem },
  mounted() {
    this.$store.dispatch("fetchRecipes");
  },
  computed: {
    ...mapGetters(["rootRecipeId"])
  }
};
</script>
