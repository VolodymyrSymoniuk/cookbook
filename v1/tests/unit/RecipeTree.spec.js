import { shallowMount, createLocalVue } from "@vue/test-utils";
import Vuex from "vuex";
import RecipeTree from "@/components/RecipeTree";
import RecipeTreeItem from "@/components/RecipeTreeItem";

const localVue = createLocalVue();
localVue.use(Vuex);

describe("RecipeTree component", () => {
  let actions, getters, wp;

  beforeEach(() => {
    getters = {
      rootRecipeId: () => "root"
    };
    actions = {
      fetchRecipes: jest.fn()
    };
    const store = new Vuex.Store({
      actions,
      getters
    });

    wp = shallowMount(RecipeTree, { store, localVue });
  });

  it("should render as expected", () => {
    expect(wp).toMatchSnapshot();
  });

  it('dispatches "fetchRecipes" on mount', () => {
    expect(actions.fetchRecipes).toHaveBeenCalled();
  });

  it("should pass right recipeId to the TreeItem", () => {
    expect(wp.findComponent(RecipeTreeItem).props("recipeId")).toBe(
      getters.rootRecipeId()
    );
  });
});
