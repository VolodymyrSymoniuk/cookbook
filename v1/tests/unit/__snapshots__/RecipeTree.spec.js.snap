// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecipeTree component should render as expected 1`] = `
<section>
  <ul>
    <recipetreeitem-stub recipeid="root"></recipetreeitem-stub>
  </ul>
</section>
`;
